streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/GraphTest.java

```diff
@@ -0,0 +1,414 @@
+        assertEquals(0, edge.residualFlow);
+
+        edges = graph.edges(2);
+        assertEquals(2, edges.size());
+
+        // No flow from 2 to 3
+        edge = edges.get(3);
+        assertEquals(0, edge.flow);
+        assertEquals(1, edge.residualFlow);
+
+        // Flow from 2 to 1 now
+        edge = edges.get(1);
+        assertEquals(1, edge.flow);
+        assertEquals(0, edge.residualFlow);
+    }
+
+    @Test
+    public void testMinCostDetectNodeNotInNegativeCycle() {
+        final Graph<Integer> graph1 = new Graph<>();
+
+        graph1.addEdge(5, 0, 1, 0, 1);
+        graph1.addEdge(5, 1, 1, 0, 1);
+
+        graph1.addEdge(0, 2, 1, 1, 0);
+        graph1.addEdge(0, 3, 1, 1, 0);
+        graph1.addEdge(0, 4, 1, 10, 1);
+
+        graph1.addEdge(1, 2, 1, 1, 0);
+        graph1.addEdge(1, 3, 1, 10, 1);
+        graph1.addEdge(1, 4, 1, 1, 0);
+
+        graph1.addEdge(2, 6, 0, 0, 0);
+        graph1.addEdge(3, 6, 1, 0, 1);
+        graph1.addEdge(4, 6, 1, 0, 1);
+
+        graph1.setSourceNode(5);
+        graph1.setSinkNode(6);
+
+        assertEquals(20, graph1.totalCost());
+
+        // In this graph, the node we found for negative cycle is 2. However 2 isn't in the negative
+        // cycle itself. Negative cycle is 1 -> 4 -> 0 -> 3 -> 1
+        graph1.solveMinCostFlow();
+        assertEquals(2, graph1.totalCost());
+
+        Map<Integer, Graph<Integer>.Edge> edges = graph1.edges(5);
+        assertEquals(getEdge(0, 1, 0, 0, 1), edges.get(0));
+        assertEquals(getEdge(1, 1, 0, 0, 1), edges.get(1));
+
+        edges = graph1.edges(0);
+        assertEquals(getEdge(2, 1, 1, 1, 0), edges.get(2));
+        assertEquals(getEdge(3, 1, 1, 0, 1), edges.get(3));
+        assertEquals(getEdge(4, 1, 10, 1, 0), edges.get(4));
+
+        edges = graph1.edges(1);
+        assertEquals(getEdge(2, 1, 1, 1, 0), edges.get(2));
+        assertEquals(getEdge(3, 1, 10, 1, 0), edges.get(3));
+        assertEquals(getEdge(4, 1, 1, 0, 1), edges.get(4));
+
+        edges = graph1.edges(2);
+        assertEquals(getEdge(6, 0, 0, 0, 0), edges.get(6));
+
+        edges = graph1.edges(3);
+        assertEquals(getEdge(6, 1, 0, 0, 1), edges.get(6));
+
+        edges = graph1.edges(4);
+        assertEquals(getEdge(6, 1, 0, 0, 1), edges.get(6));
+    }
+
+    @Test
+    public void testDeterministic() {
+        final List<TestEdge> edgeList = new ArrayList<>();
+        edgeList.add(new TestEdge(0, 1, 1, 2, 1));
+        edgeList.add(new TestEdge(0, 2, 1, 1, 0));
+        edgeList.add(new TestEdge(0, 3, 1, 1, 0));
+        edgeList.add(new TestEdge(0, 4, 1, 1, 0));
+        edgeList.add(new TestEdge(1, 5, 1, 1, 1));
+        edgeList.add(new TestEdge(2, 5, 1, 1, 0));
+        edgeList.add(new TestEdge(3, 5, 1, 1, 0));
+        edgeList.add(new TestEdge(4, 5, 1, 1, 0));
+
+        // Test no matter the order of adding edges, min cost flow flows from 0 to 2 and then from 2 to 5
+        for (int i = 0; i < 10; i++) {
```

Comment:
Why do we need to test this 10 times? Given that the test runs for each PR on nightly builds, it seems sufficient to just run it once?

Reply:
Want to test it more if someone trigger this manually. I feel 10 times won't hurt much, but I'm fine switching to 1 if you feel strongly 😄

Reply:
Test should run fast. Don't feel strongly about it.